doc/design/drafts/dynamic_size_types.md

```diff
@@ -0,0 +1,562 @@
+       |  // same API as current cxx::vector         |
+       |                                             |
+       |  # uint8_t m_data[Capacity]                 |
+       |  # RangeAllocator m_allocator               |
+       +---------------------------------------------+
+```
+All non stack based versions of cxx containers will not have constructors similar
+to the STL. The only provided constructor has one argument which specifies the
+type of allocator one would like to use for that container. The default constructor
+shall construct a vector which has a capacity of 0.
+
+The constructors and operations defined in (#1) are added to support implicit
+conversion from the generic `cxx::vector`. It makes the following operations possible
+```cpp
+void f(const vector<int, 20> &a) {}
+
+vector<int> v1;
+vector<int, 20> v2;
+
+// this works thanks to implicit conversions from vector<int>
+f(v1);
+v2 = v1;
+```
+We can use `cxx::Expects` to ensure that the size of `v1` is smaller or equal
+than the capacity `v2`. Later we can overload the copy operation with a method
+like `perform_copy` which returns the boolean `true` when the copy operation
+was successful and `false` when the size of `v1` exceeded the capacity of `v2`.
+
+Since every stack based vector is a child of the more generic vector we can, with
+the help of (#1), also assign and use stack based vectors of different sizes
+to each other or use them in functions. This allows us to write code like this:
+```cpp
+void f(const vector<int, 20> &a) {}
+
+vector<int, 10> v1;
+vector<int, 20> v2;
+
+// this works since vector<int, 10> is also a vector<int> and can be converted
+v2 = v1;
+v1 = v2; // this can fail when v1 capacity is smaller then v2.size, previously 
+         // it was only possible when v1.capacity < v2.capacity
+f(v1);
+f(v2);
+```
+
+The only limitation is functions which require a stack based vector of a 
+specific size as non const reference. This is usally the case when some function
+would like to set the contents of a vector.
+```cpp
+void setVectorContents(vector<int, 20> &a) {}
+
+vector<int, 10> v1;
+vector<int> v2;
+
+// both calls will result in a compile time warning
+setVectorContents(v1);
+setVectorContents(v2);
+```
+
+### Usage
+
+```cpp
+cxx::vector<int, 20> a;
+a.emplace_back(123); // can be used without a preceding reserve call
+
+cxx::vector<int> c(heapAllocator);
+// reserve required
+c.reserve(42)
+     .and_then([&]{ c.emplace_back(891); })
+     .or_else([](auto & error){ /* error handling */});
+
+auto sample = publisher.loan();
+if ( sample->sharedMemoryVector.reserve(123).has_error() {
+   /* error handling */
+   return;
+}
+sample->sharedMemoryVector.emplace_back(someValue);
+```
+
+
+Copy and move operations should be possible without the developer knowing
+what kind of underlying allocator is used. Additionally, it should be as efficient
+as possible. That means the move operation should be a real move operation
```

Comment:
We could also define that move only works if source and destination allocator are the same (and maybe other conditions hold), otherwise we could always fall back to a copy.

Another approach is simply that the destiantion takes over the memory and allocator of the source and discards its own. If the allocator is not owned by the vector (but just a fancy pointer to it), this should be well defined and cheap, at least for vector.

The source of the move will not need allocator nor data (it should not be accessed anymore, but we can make it move aware for added safety at the cost of some runtime checks whether the allocator is valid/set).